File: lib/src/models/library_document.dart
import 'document_format.dart';

/// A document stored in the app's library (a PDF or a reflowable book such as
/// EPUB, MOBI, FB2, TXT, HTML, Markdown or RTF).
///
/// The file itself is copied into the app's documents directory on import so
/// the library is self-contained (and keeps working if the original file is
/// moved or deleted).
class LibraryDocument {
  const LibraryDocument({
    required this.id,
    required this.title,
    required this.filePath,
    this.originalPath,
    this.pageCount = 0,
    required this.addedAt,
    this.lastOpenedAt,
    this.lastPage = 1,
    this.viewMatrix,
    this.folderId,
    this.position = 0,
  });

  final int id;

  /// Display title (defaults to the file name without extension).
  final String title;

  /// Absolute path to the managed copy inside the app library directory.
  final String filePath;

  /// Where the file was imported from, for reference.
  final String? originalPath;

  /// Number of pages, or 0 if not yet determined.
  final int pageCount;

  final DateTime addedAt;
  final DateTime? lastOpenedAt;

  /// 1-based page the reader was last on (reading position).
  final int lastPage;

  /// Serialized pdfrx view matrix (16 comma-separated doubles) capturing the
  /// exact scroll position and zoom; null until the document has been viewed.
  final String? viewMatrix;

  /// Id of the folder this document belongs to, or null for the library root.
  final int? folderId;

  /// Order of this document among the items of its folder (or the root).
  final int position;

  /// The document's format, derived from its file extension.
  DocumentFormat get format => documentFormatForPath(filePath);

  LibraryDocument copyWith({
    int? id,
    String? title,
    String? filePath,
    Object? originalPath = _sentinel,
    int? pageCount,
    DateTime? addedAt,
    Object? lastOpenedAt = _sentinel,
    int? lastPage,
    Object? viewMatrix = _sentinel,
    Object? folderId = _sentinel,
    int? position,
  }) {
    return LibraryDocument(
      id: id ?? this.id,
      title: title ?? this.title,
      filePath: filePath ?? this.filePath,
      originalPath: originalPath == _sentinel
          ? this.originalPath
          : originalPath as String?,
      pageCount: pageCount ?? this.pageCount,
      addedAt: addedAt ?? this.addedAt,
      lastOpenedAt: lastOpenedAt == _sentinel
          ? this.lastOpenedAt
          : lastOpenedAt as DateTime?,
      lastPage: lastPage ?? this.lastPage,
      viewMatrix: viewMatrix == _sentinel
          ? this.viewMatrix
          : viewMatrix as String?,
      folderId: folderId == _sentinel ? this.folderId : folderId as int?,
      position: position ?? this.position,
    );
  }

  static const Object _sentinel = Object();
}

File: ios-native/Eng/Models/LibraryDocument.swift
import Foundation

/// A PDF imported into the app-managed library. The file itself is a copy kept
/// under the app's Documents/library folder; `fileName` is its name there.
struct LibraryDocument: Identifiable, Equatable {
    var id: Int64
    var title: String
    /// File name inside the managed `library/` directory (not an absolute path,
    /// so it survives the container path changing between installs).
    var fileName: String
    /// The path the file was imported from, for reference/display.
    var originalPath: String?
    var pageCount: Int
    var addedAt: Date
    var lastOpenedAt: Date?
    /// 1-based page the reader should resume on (fallback when `viewState` is absent).
    var lastPage: Int = 1
    /// Exact reading position + zoom, restored verbatim on the next open.
    var viewState: DocumentViewState? = nil

    /// Absolute URL of the managed copy, resolved against the current container.
    var fileURL: URL { AppPaths.libraryDirectory.appendingPathComponent(fileName) }
}

/// The reader's exact view of a document: the current page, the top-left point of
/// the visible area in PDF page space (what `PDFView.currentDestination` reports),
/// and the absolute zoom (`scaleFactor`). Serialized as JSON into the documents
/// table's `view_matrix` column (the same column the Flutter app uses).
struct DocumentViewState: Codable, Equatable {
    /// Format version. v2 = point computed via `convert(_:to:)` (true page space).
    /// The key is REQUIRED when decoding, so states saved by the first build —
    /// whose `currentDestination`-derived points restored about a page off — fail
    /// to decode and fall back to the plain last-page resume.
    var v: Int = 2
    /// 0-based page index.
    var pageIndex: Int
    /// Top-left of the visible area, in page space (origin bottom-left).
    var x: Double
    var y: Double
    /// `PDFView.scaleFactor` at the time of saving.
    var zoom: Double

    var json: String? {
        guard let data = try? JSONEncoder().encode(self) else { return nil }
        return String(decoding: data, as: UTF8.self)
    }

    static func from(json: String?) -> DocumentViewState? {
        guard let json else { return nil }
        return try? JSONDecoder().decode(DocumentViewState.self, from: Data(json.utf8))
    }
}

/// Resolves the app's on-disk locations. Everything lives under the app-support
/// dir: the SQLite database plus a `library/` folder holding copies of imports.
enum AppPaths {
    static var supportDirectory: URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        try? FileManager.default.createDirectory(at: base, withIntermediateDirectories: true)
        return base
    }

    static var libraryDirectory: URL {
        let dir = supportDirectory.appendingPathComponent("library", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    static var databaseURL: URL { supportDirectory.appendingPathComponent("eng.db") }
}
